CORE-10576 Trigger member list distribution when suspending and activating a member

williamvigorr3 opened the pull request:
Trigger member list distribution from the rest worker after updating the database when suspending or activating a member. Additionally change the db transactions to remove and add a notary from the group parameters if a member is suspended/activated. Also changed the distribution protocol inside DistributeMemberInfoAction to only send their own member info to a suspended member (so information is not leaked about the state of the network).  Also changed the `Member` and MGM side of the Synchronisation service to reflect this. Requires API changes: https://github.com/corda/corda-api/pull/1041

williamvigorr3 committed:
CORE-10576 Trigger member list distribution when suspending and activating a member

williamvigorr3 committed:
Fix broken test

williamvigorr3 committed:
Revert test changes

williamvigorr3 committed:
Change distribution so that Suspended members only receive there own member info

williamvigorr3 committed:
Enable suspension/activation rest requests.

nikinagy commented on libs/membership/membership-common/src/main/kotlin/net/corda/membership/lib/InternalGroupParametersUtils.kt:
```diff
@@ -0,0 +1,31 @@
+    val signed = this as? SignedGroupParameters
+    return PersistentGroupParameters(
+        owner.toAvro(),
+        net.corda.data.membership.SignedGroupParameters(
```
Maybe we can introduce a type alias for this and call it "SignedGroupParametersAvro" to make the code a bit prettier? But it's just an idea, we can leave it as is if we prefer that way

williamvigorr3 replied:
I imported it as `AvroSignedGroupParameters`.

nikinagy commented on components/membership/membership-persistence-client/src/main/kotlin/net/corda/membership/persistence/client/MembershipPersistenceClient.kt:
```diff
@@ -283,7 +283,7 @@ interface MembershipPersistenceClient : Lifecycle {
         serialNumber: Long?,
         reason: String?,
-    ): MembershipPersistenceResult<PersistentMemberInfo>
+    ): MembershipPersistenceResult<Pair<PersistentMemberInfo, InternalGroupParameters?>>
```
Modify the kdocs's return part as well

nikinagy commented on components/membership/membership-persistence-client/src/main/kotlin/net/corda/membership/persistence/client/MembershipPersistenceClient.kt:
```diff
@@ -300,7 +300,7 @@ interface MembershipPersistenceClient : Lifecycle {
         serialNumber: Long?,
         reason: String?,
-    ): MembershipPersistenceResult<PersistentMemberInfo>
+    ): MembershipPersistenceResult<Pair<PersistentMemberInfo, InternalGroupParameters?>>
```
Modify the kdocs's return part as well

nikinagy commented on components/membership/membership-client-impl/src/main/kotlin/net/corda/membership/impl/client/MGMResourceClientImpl.kt:
```diff
@@ -573,17 +588,38 @@ class MGMResourceClientImpl @Activate constructor(
         }
 
-        private fun publishMemberInfo(memberInfo: PersistentMemberInfo, mgm: String, member: String) {
+        private fun publishMemberInfoAndRequestDistribution(
```
I am not a fan of the name as here we can publish group params too, a bit misleading.

williamvigorr3 replied:
I renamed this `publishSuspensionActivationRecords`.

nikinagy commented on components/membership/membership-client-impl/src/main/kotlin/net/corda/membership/impl/client/MGMResourceClientImpl.kt:
```diff
@@ -573,17 +588,38 @@ class MGMResourceClientImpl @Activate constructor(
+            mgmShortHash: String,
+        ) {
+            val serialNumber = memberInfoFactory.create(memberInfo).serial
+            val publisher = coordinator.getManagedResource<Publisher>(PUBLISHER_RESOURCE_NAME)
```
is there a good reason why we moved these out from the try-catch?

nikinagy replied:
ok don't mind this comment I think its because we are just catching CordaMessageAPIIntermittentException anyway.

williamvigorr3 replied:
Yes.

nikinagy commented on components/membership/membership-client-impl/src/test/kotlin/net/corda/membership/impl/client/MGMResourceClientTest.kt:
```diff
@@ -1578,7 +1618,7 @@ class MGMResourceClientTest {
             val mockMemberInfo = mock<PersistentMemberInfo>()
             whenever(membershipPersistenceClient.activateMember(eq(holdingIdentity), eq(memberName), any(), any()))
-                .doReturn(MembershipPersistenceResult.Success(mockMemberInfo))
+                .doReturn(MembershipPersistenceResult.Success(mockMemberInfo to null))
```
Can we add a test where the group params won't be null?

williamvigorr3 replied:
I have added `suspendMember should publish the updated group parameters, the updated member info and request distribution` and `activateMember should publish the updated group parameters, the updated member info and request distribution`.

nikinagy commented on components/membership/membership-persistence-client-impl/src/test/kotlin/net/corda/membership/impl/persistence/client/MembershipPersistenceClientImplTest.kt:
```diff
@@ -1182,7 +1182,7 @@ class MembershipPersistenceClientImplTest {
         @Test
         fun `suspendMember returns the correct result`() {
-            mockPersistenceResponse(SuspendMemberResponse(mock()))
+            mockPersistenceResponse(SuspendMemberResponse(mock(), mock()))
```
maybe verify that groupParamatersFactory.create has been called?

williamvigorr3 replied:
I split this test (and the same for activateMember) into two cases (one where the group parameters is null, one where it is not). I added the verifies as suggested.

nikinagy commented on components/membership/membership-persistence-service-impl/src/integrationTest/kotlin/net/corda/membership/impl/persistence/service/MembershipPersistenceTest.kt:
```diff
@@ -1404,7 +1404,7 @@ class MembershipPersistenceTest {
         val suspended1 = membershipPersistenceClientWrapper.activateMember(
             viewOwningHoldingIdentity, member1, 1L, "test-reason"
-        ).getOrThrow()
+        ).getOrThrow().first
```
can we have a test for notary too? and see what is returned with the group params

nikinagy commented on components/membership/membership-persistence-service-impl/src/main/kotlin/net/corda/membership/impl/persistence/service/handler/ActivateMemberHandler.kt:
```diff
@@ -1,24 +1,59 @@
-            MEMBER_STATUS_SUSPENDED,
-            MEMBER_STATUS_ACTIVE
-        )
+        val (updatedMemberInfo, updatedGroupParameters) = transaction(context.holdingIdentity.toCorda().shortHash) { em ->
```
Just a nit, but adding log messages to handlers would make life easier in the long run if something goes wrong.

nikinagy commented on components/membership/membership-persistence-service-impl/src/main/kotlin/net/corda/membership/impl/persistence/service/handler/SuspendMemberHandler.kt:
```diff
@@ -1,24 +1,140 @@
+            val currentMgmContext = keyValuePairListDeserializer.deserialize(currentMemberInfo.mgmContext)
+                ?: throw MembershipPersistenceException("Failed to deserialize the MGM-provided context.")
+            PersistentMemberInfo(
+                HoldingIdentity(request.suspendedMember, context.holdingIdentity.groupId),
```
I am a bit lost here: why the view owning member is the suspended member? Same goes in ActivateMemberHandler. What are we even doing with these PersistentMemberInfos here? it seems unused

nikinagy replied:
I think here and in the ActivateMemberHandler these were left by mistake

williamvigorr3 replied:
Yes I deleted this now.

nikinagy commented on components/membership/membership-persistence-service-impl/src/test/kotlin/net/corda/membership/impl/persistence/service/handler/ActivateMemberHandlerTest.kt:
```diff
@@ -87,12 +89,20 @@ class ActivateMemberHandlerTest {
             createAvroSerializer<KeyValuePairList>(any())
         } doReturn keyValuePairListSerializer
     }
+    private val memberInfo = mock<MemberInfo> {
```
test with notary too

nikinagy commented on components/membership/membership-persistence-service-impl/src/test/kotlin/net/corda/membership/impl/persistence/service/handler/SuspendMemberHandlerTest.kt:
```diff
@@ -87,12 +89,21 @@ class SuspendMemberHandlerTest {
             createAvroSerializer<KeyValuePairList>(any())
         } doReturn keyValuePairListSerializer
     }
+    private val memberInfo = mock<MemberInfo> {
```
test with notary too and always check the test coverage on unit tests

nikinagy commented on components/membership/membership-service-impl/src/main/kotlin/net/corda/membership/service/impl/actions/DistributeMemberInfoActionHandler.kt:
```diff
@@ -110,12 +110,20 @@ class DistributeMemberInfoActionHandler(
+        val allActiveMembersExcludingMgm = allActiveMembers.filterNot { it.isMgm }
+        //If the updated member is suspended then we only send it's own member info to itself (so it can tell it has been suspended).
+        val membersToDistributeToUpdatedMember = if (updatedMemberInfo.status == MEMBER_STATUS_SUSPENDED) {
+            listOf(updatedMemberInfo)
```
I am a bit a worried if this will work with the hash checks in place. I think the suspended member will keep asking for syncs from MGM, because the hash is not matching the expected hash.

williamvigorr3 replied:
😟

nikinagy replied:
I am still worried this won't work as we imagined. Did we double-checked this?

williamvigorr3 replied:
I am going to do that now (when I run the e2e tests).

nikinagy replied:
I think it will settle eventually by the synchronisation protocol due to the fact the hashes won't match when the member receives the data package. So it will just do an extra round of sync.

charlieR3 commented on libs/membership/membership-common/src/main/kotlin/net/corda/membership/lib/GroupParametersNotaryUpdater.kt:
```diff
@@ -0,0 +1,209 @@
+    }
+    private val logger = LoggerFactory.getLogger(this::class.java)
+
+    fun addNewNotaryService(
```
Should `addNewNotaryService` be removed from `GroupParametersUtils` now after being moved here?

williamvigorr3 replied:
I am going to delete the `GroupParametersUtils`.

charlieR3 commented on libs/membership/membership-common/src/main/kotlin/net/corda/membership/lib/GroupParametersNotaryUpdater.kt:
```diff
@@ -0,0 +1,209 @@
+        return newEpoch to KeyValuePairList(parametersWithUpdatedEpoch + newService)
+    }
+
+    fun updateExistingNotaryService(
```
Same here, should `updateExistingNotaryService` be removed from `GroupParametersUtils` now after being moved here?

charlieR3 commented on libs/membership/membership-common/src/test/kotlin/net/corda/membership/lib/GroupParametersNotaryUpdaterTest.kt:
```diff
@@ -0,0 +1,247 @@
+        )
+    }
+
+    //Todo test exceptions.
```
TODO comment. Does it need to be removed or are there tests remaining to add?

williamvigorr3 replied:
I need to add the tests and probably rework the `AddNotaryToGroupParametersHandler` to be a purer unit test (coverage of the exceptions will move from their to here).

charlieR3 commented on components/membership/membership-rest/src/main/kotlin/net/corda/membership/rest/v1/MGMRestResource.kt:
```diff
@@ -461,9 +461,9 @@ interface MGMRestResource : RestResource {
-//    @HttpPOST(
-//        path = "{holdingIdentityShortHash}/activate"
-//    )
+    @HttpPOST(
```
Does this need an update to swaggerBaseline?

williamvigorr3 replied:
Yes I forgot to do that.

charlieR3 commented on components/membership/membership-persistence-service-impl/src/main/kotlin/net/corda/membership/impl/persistence/service/handler/BaseSuspensionActivationHandler.kt:
```diff
@@ -14,25 +15,96 @@ import net.corda.membership.lib.MemberInfoExtension.Companion.STATUS
+                        "MemberInfo for member '$memberName'.")
+            }
+        }
+        require(member.status == expectedStatus) {
```
Maybe if the member is already in the expected status we can just skip the update instead of throwing an exception

charlieR3 commented on components/membership/membership-persistence-service-impl/src/main/kotlin/net/corda/membership/impl/persistence/service/handler/SuspendMemberHandler.kt:
```diff
@@ -1,24 +1,140 @@
+            )
+            val currentMgmContext = keyValuePairListDeserializer.deserialize(currentMemberInfo.mgmContext)
+                ?: throw MembershipPersistenceException("Failed to deserialize the MGM-provided context.")
+            PersistentMemberInfo(
```
Unused?

williamvigorr3 replied:
I remove this.

charlieR3 commented on components/membership/membership-persistence-service-impl/src/main/kotlin/net/corda/membership/impl/persistence/service/handler/SuspendMemberHandler.kt:
```diff
@@ -1,24 +1,140 @@
+                currentMgmContext,
+                MEMBER_STATUS_SUSPENDED
+            )
+            val updatedGroupParameters = if (memberInfoFactory.create(updatedMemberInfo).notaryDetails != null) {
```
There's an `isNotary` function is MemberInfoExtension which could be used to check if it's a notary without the additional notary details parsing if the notary details aren't needed at this point.

charlieR3 commented on components/membership/membership-persistence-service-impl/src/main/kotlin/net/corda/membership/impl/persistence/service/handler/SuspendMemberHandler.kt:
```diff
@@ -1,24 +1,140 @@
+        val otherMembersOfSameNotaryService = members.filter {
+            it.notaryDetails?.serviceName.toString() == notaryServiceName &&
+                    it.name != notaryInfo.name &&
+                    it.status == MEMBER_STATUS_ACTIVE
```
Could the status be part of the DB query for member infos to reduce the number of member infos that need to be parsed?

charlieR3 replied:
Maybe even the name != notaryInfo.name could be added to the DB query. That way we could probably have `.mapNotNull { it.notaryDetails }.filter {serviceName.toString() == notaryServiceName }`

charlieR3 replied:
These are all kind of optimizations thinking ahead to large network support and trying to reduce the amount of times we need to parse `notaryDetails`

williamvigorr3 replied:
Ideally we could check if the member was a notary in the query (but we can't because the member context has been serialised). 

I think even with the optimisations you have suggested we could potentially be querying most of the member list here in ActivateMamberHandler and in AddNotaryToGroupParametersHandler (most members will likely be `ACTIVE` and `name != notaryInfo.name` will reduce the number of members by one).

charlieR3 commented on components/membership/membership-persistence-service-impl/src/main/kotlin/net/corda/membership/impl/persistence/service/handler/SuspendMemberHandler.kt:
```diff
@@ -1,24 +1,140 @@
+        val memberQuery = memberQueryBuilder.select(memberQueryBuilder.from(MemberInfoEntity::class.java))
+        val members = em.createQuery(memberQuery)
+            .setLockMode(LockModeType.PESSIMISTIC_WRITE)
+            .resultList.map {
```
Maybe something to look at during large network support (so I don't mind not blocking the PR on this) but `resultStream` could be better for querying the entire member list if it's a very large list.

charlieR3 commented on components/membership/membership-persistence-service-impl/src/main/kotlin/net/corda/membership/impl/persistence/service/handler/AddNotaryToGroupParametersHandler.kt:
```diff
@@ -52,84 +56,77 @@ internal class AddNotaryToGroupParametersHandler(
+            val memberQuery = memberQueryBuilder.select(memberQueryBuilder.from(MemberInfoEntity::class.java))
+            val members = em.createQuery(memberQuery)
+                .setLockMode(LockModeType.PESSIMISTIC_WRITE)
+                .resultList.map {
```
Same comment here in that we could stream the results of querying the full member list, which could be quite large. But I'm willing to cover it later if necessary.

williamvigorr3 replied:
I was thinking about this I am not sure we need to query all the members here (I decided not to change the behaviour as a part of this PR). As we want to union the set of notary versions for each member of the notary service. We could just union with the set of versions already in the group parameters, each time we update it (as union is associative), instead of recomputing the whole thing. Unless I missed something.

The same is not true for the `SuspendMemberHandler` (we need to recalculate the union of all the other not suspended members if we suspend a member of the notary service).

charlieR3 replied:
That's a good point. I think that could work for adding a notary but not removing as you said

charlieR3 commented on components/membership/membership-persistence-service-impl/src/main/kotlin/net/corda/membership/impl/persistence/service/handler/ActivateMemberHandler.kt:
```diff
@@ -1,24 +1,59 @@
+            )
+            val currentMgmContext = keyValuePairListDeserializer.deserialize(currentMemberInfo.mgmContext)
+                ?: throw MembershipPersistenceException("Failed to deserialize the MGM-provided context.")
+            PersistentMemberInfo(
```
Unused?

charlieR3 commented on components/membership/membership-persistence-service-impl/src/main/kotlin/net/corda/membership/impl/persistence/service/handler/ActivateMemberHandler.kt:
```diff
@@ -1,24 +1,59 @@
+                currentMgmContext,
+                MEMBER_STATUS_ACTIVE
+            )
+            val updatedGroupParameters = if (memberInfoFactory.create(updatedMemberInfo).notaryDetails != null) {
```
```suggestion
            val updatedGroupParameters = if (memberInfoFactory.create(updatedMemberInfo).isNotary()) {
```

charlieR3 commented on components/membership/membership-client-impl/src/main/kotlin/net/corda/membership/impl/client/MGMResourceClientImpl.kt:
```diff
@@ -573,17 +588,38 @@ class MGMResourceClientImpl @Activate constructor(
+        ) {
+            val serialNumber = memberInfoFactory.create(memberInfo).serial
+            val publisher = coordinator.getManagedResource<Publisher>(PUBLISHER_RESOURCE_NAME)
+            val recordForGroupParameters = groupParameters?.let {
```
This seems incorrect. Is this publishing the group parameters for another member? Other members should receive the updated group parameters through the data distribution protocol and the MGM shouldn't attempt to write the members view of the group parameters to kafka (especially since they probably won't be on the same cluster)

williamvigorr3 replied:
Good spot. This should publish the MGM's view of the group parameters not the members.

charlieR3 replied:
You might be able to use `GroupParametersWriterService` for that if it reduces duplication

williamvigorr3 replied:
I fixed this. I didn't use the `GroupParametersWriterService` as so we can ignore errors if writing to Kafka failed.

williamvigorr3 committed:
Update swagger baseline

williamvigorr3 committed:
Merge remote-tracking branch 'origin/release/os/5.0' into WillV/CORE-10576-Trigger-Member-list-distribution-when-suspending-and-activating-a-member

williamvigorr3 committed:
Respond to review comments

williamvigorr3 committed:
Minor fixes based on code review

williamvigorr3 committed:
Improve tests

williamvigorr3 committed:
Respond to more review comments + improve unit tests

williamvigorr3 committed:
Merge remote-tracking branch 'origin/release/os/5.0' into WillV/CORE-10576-Trigger-Member-list-distribution-when-suspending-and-activating-a-member

williamvigorr3 committed:
Change MemberSynchronisationServiceImpl + MgmSynchronisationServiceImpl

To work in the same way as the DistributeMemberInfoActionHandler.
On the Memberside we only hash our own member info, if we have
been suspended. On the MGM side only distribute a member list with
our own Member info. Also delete GroupParametersUtils.kt (GroupParametersNotaryUpdater
is used instead. Improve some of the unit tests.

dimosr commented on components/membership/synchronisation-impl/src/main/kotlin/net/corda/membership/impl/synchronisation/MemberSynchronisationServiceImpl.kt:
```diff
@@ -333,9 +336,15 @@ class MemberSynchronisationServiceImpl internal constructor(
-                    val expectedHash = merkleTreeGenerator.generateTree(allMembers.values).root
+                    val knownMembers = groupReader.lookup(MembershipStatusFilter.ACTIVE_OR_SUSPENDED)
+                        .filter { !it.isMgm }.associateBy { it.id }
+                    val updatedViewOwner = updateMembersInfo[viewOwningMember.shortHash.value]
```
Is it safer to perform a lookup on `knownMembers` as well if not found in `updateMembersInfo` to capture cases where we might be sent other members info, but we might still be suspended?

charlieR3 replied:
Posted on slack, but will comment here too for visibility. I think we don't have a case currently where a member should be sent other member infos while still suspended so I would say the way it's implemented here is OK.

charlieR3 replied:
Although, thinking again, double checking against the member list would protect against messages outside of what is expected like replays or some mistake on the MGM side so I'm thinking now it could be wise to check against known members also for our own suspended status.

dimosr replied:
[Discussed offline with everyone and agreed to do that change, but as a follow-up along with all the other pending comments in order to get this functionality in for the next RC/HC]

dimosr commented on components/membership/synchronisation-impl/src/main/kotlin/net/corda/membership/impl/synchronisation/MgmSynchronisationServiceImpl.kt:
```diff
@@ -213,16 +216,20 @@ class MgmSynchronisationServiceImpl internal constructor(
                 ?: throw CordaRuntimeException("Requester $requesterName $IDENTITY_EX_MESSAGE")
             // we don't want to include the MGM in the data package since MGM information comes from the group policy
-            val allMembers = groupReader.lookup().filterNot { it.holdingIdentity == mgm.toCorda() }
+            val allMembersActiveMemberExcludingMgm = groupReader.lookup().filterNot { it.holdingIdentity == mgm.toCorda() }
```
Don't we want to send both ACTIVE and SUSPENDED members here now? Seems like this lookup will only return ACTIVE ones by default.

williamvigorr3 replied:
Yes. I did this in b0391ac.

williamvigorr3 committed:
Only distribute own member info to suspended members in the MgmSynchronisationServiceImpl

williamvigorr3 commented:
List of things I think needs to be done (in my opinion):
1. Add coverage of notary suspension/activation to `SuspendMemberHandlerTest` and `ActivateMemberHandlerTest`.
2. Add more test coverage to MembershipPersistenceTest (there is some coverage of notary suspension I could add more (although these tests are expensive and this will be covered by unit tests) I need to test notary activations.
3. Unit test changes to the MgmSynchronisation service.

Things which could be done in a follow up (in my opinion):
4. Tidy SuspensionActivationEntityOperationsTest (assuming there is enough coverage here).
5. Move coverage of exceptions from the  `AddNotaryToGroupParameterHandlerTest` to the `GroupParametersNotaryUpdaterTest`. To make purer unit tests.

dimosr reviewed (commented):
I only had a look at the changes in the sync process, posting a couple of questions

williamvigorr3 committed:
Merge remote-tracking branch 'origin/release/os/5.0' into WillV/CORE-10576-Trigger-Member-list-distribution-when-suspending-and-activating-a-member

charlieR3 commented on components/membership/membership-persistence-service-impl/src/test/kotlin/net/corda/membership/impl/persistence/service/handler/SuspensionActivationEntityOperationsTest.kt:
```diff
@@ -0,0 +1,382 @@
+        }
+    }
+
+//    @Test
```
Tests should be uncommented or removed before merging

williamvigorr3 replied:
I will remove this for now, we need to cover some of this in `SuspendMemberHandlerTest` / `ActivateMemberHandlerTest`.

charlieR3 reviewed (commented):
No additional comments (other than one minor one about commented tests)

williamvigorr3 committed:
Distribute Suspended members to active members in MgmSynchronisationService

williamvigorr3 committed:
Merge remote-tracking branch 'origin/release/os/5.0' into WillV/CORE-10576-Trigger-Member-list-distribution-when-suspending-and-activating-a-member

williamvigorr3 committed:
Remove tests + bump version

nikinagy reviewed (dismissed):
LGTM

vkolomeyko commented on processors/rest-processor/src/integrationTest/resources/swaggerBaseline.json:
```diff
@@ -5088,6 +5170,29 @@
+            "example" : "string"
+          }
+        },
+        "description" : "Parameters for activating a suspended member."
```
```suggestion
        "description" : "Parameters for activating or suspending a member."
```

williamvigorr3 replied:
I will fix that (once I get the merge lock).

williamvigorr3 replied:
I change it too `Parameters for suspending or activating a member.` to be consistent with the KDOC on `SuspensionActivationParameters`.

vkolomeyko reviewed (dismissed):
LGTM

Minor comment posted

williamvigorr3 committed:
Merge remote-tracking branch 'origin/release/os/5.0' into WillV/CORE-10576-Trigger-Member-list-distribution-when-suspending-and-activating-a-member

williamvigorr3 committed:
Update description of SuspensionActivationParameters

vkolomeyko reviewed (approved):
LGTM
